src/sentry/data_secrecy/data_secrecy_logic.py

```diff
@@ -29,3 +34,121 @@ def should_allow_superuser_access(
+    :return: True if a superuser session is allowed for the organization, False otherwise.
+
+    Note: This method is not currently used. It will be rolled out when rest of Data Secrecy V2 is ready.
+    It is an in-place replacement for should_allow_superuser_access.
+    """
+
+    if settings.SENTRY_SELF_HOSTED:
+        return True
+
+    organization: Organization | RpcOrganization
+    if isinstance(organization_context, RpcUserOrganizationContext):
+        organization = organization_context.organization
+    else:
+        organization = organization_context
+
+    # If organization does not have data-secrecy-v2 feature, allow superuser access
+    if not features.has("organizations:data-secrecy-v2", organization):
+        return True
+
+    # If organization's prevent_superuser_access bitflag is False, allow superuser access
+    if not organization.flags.prevent_superuser_access:
+        return True
+
+    # If there is an active grant, allow superuser access
+    if data_access_grant_exists(organization.id):
+        return True
+
+    return False
+
+
+def _get_cached_grant_status(organization_id: int) -> bool | None:
+    """
+    Retrieve cached grant status for an organization.
+
+    :param organization_id: The ID of the organization to get the cached grant status for.
+    :return:
+        - False if negative cache hit.
+        - True if cached data is valid and not expired.
+        - None if not cached.
+    """
+    cache_key = CACHE_KEY_PATTERN.format(organization_id=organization_id)
+    cached_data = cache.get(cache_key)
+
+    if cached_data:
+        # Check if it's negative cache
+        if cached_data == NEGATIVE_CACHE_VALUE:
+            return False
+
+        # Verify the cached data hasn't logically expired
+        access_end = cached_data["access_end"]
+        if access_end > datetime.now(timezone.utc):
+            return True
+        else:
+            # Cached data is stale, remove it
+            cache.delete(cache_key)
+
+    return None
+
+
+def data_access_grant_exists(organization_id: int) -> bool:
+    """
+    Determines if a data access grant exists for an organization.
+    Tries cache first, calculates and caches if needed.
+
+    :param organization_id: The ID of the organization to check.
+    :return: True if a data access grant exists for the organization, False otherwise.
+    """
+    # Try cache first
+    cached_status = _get_cached_grant_status(organization_id)
+    if cached_status is not None:
+        return cached_status
+
+    # Cache miss or expired, calculate fresh
+    grant_status = data_access_grant_service.get_effective_grant_status(
+        organization_id=organization_id
+    )
+
+    if grant_status:
+        # Calculate TTL first to avoid race condition where grant expires between checks
+        current_time = datetime.now(timezone.utc)
+        ttl_seconds = int((grant_status.access_end - current_time).total_seconds())
```

Comment:
This type of logic could live on a dataclass we rehydrate from the cache to clean things up. Likewise, this might give us an opportunity to combine this with the negative cache object in a way that makes the two more cohesive.